<a href="https://colab.research.google.com/github/Psychraft/Project_EPAM/blob/main/EPAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Victor Oscar Bernal Alpizar
Student ID 4156
  
Final Project Big Data Analysis

                                      Professor: Piotr Radkowski 


In [ ]:
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():

  print('User uploaded file "{name}" with length {length} bytes'.format(

name=fn, length=len(uploaded[fn])))

Saving cardio_train.csv to cardio_train.csv
User uploaded file "cardio_train.csv" with length 2941524 bytes


# New Section

In [ ]:
import pandas as pd #Data processing
import pandas.util.testing as tm
import numpy as np
from scipy import stats

# Data visualization  
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn import metrics
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import os # accessing directory structure
from mpl_toolkits.mplot3d import Axes3D
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt # plotting
import numpy as np # linear algebra

from sklearn.preprocessing import RobustScaler, label_binarize
from sklearn.model_selection import train_test_split, GridSearchCV

# decision tree, logistic
from sklearn.tree import DecisionTreeClassifier, export_graphviz
import graphviz
from sklearn.linear_model import LogisticRegressionCV

import statsmodels.api as sm       
from statsmodels.formula.api import ols, logit
import statsmodels.formula.api as smf 

# VIF computation
from statsmodels.stats.outliers_influence import variance_inflation_factor

# ensemble methods, MLP
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.svm import LinearSVC, SVC
from sklearn.neural_network import MLPClassifier

from itertools import product

from sklearn.metrics import jaccard_score, adjusted_rand_score, silhouette_score, calinski_harabasz_score, roc_curve, auc, accuracy_score, classification_report, confusion_matrix
from sklearn.metrics.cluster import contingency_matrix




In [ ]:
#To read the dataset that was uploaded in colab 
Cardio_disease = pd.read_csv (r'/content/cardio_train.csv')

NameError: ignored

In [ ]:
Cardio_disease = pd.read

NameError: ignored

In [ ]:
Cardio_disease


First step to be done is the inspection of the files and to check if the data was neither corrupted nor damaged which I made sure this wont happen. Also considering the semicolon as the separator so that all observations can be readeable.

In [ ]:
#Dataset needs to be delimited with semicolon
Cardio_disease2 = pd.read_csv (r'/content/cardio_train.csv', delimiter=';')


In [ ]:
print(Cardio_disease2)

          id    age  gender  height  weight  ...  gluc  smoke  alco  active  cardio
0          0  18393       2     168    62.0  ...     1      0     0       1       0
1          1  20228       1     156    85.0  ...     1      0     0       1       1
2          2  18857       1     165    64.0  ...     1      0     0       0       1
3          3  17623       2     169    82.0  ...     1      0     0       1       1
4          4  17474       1     156    56.0  ...     1      0     0       0       0
...      ...    ...     ...     ...     ...  ...   ...    ...   ...     ...     ...
69995  99993  19240       2     168    76.0  ...     1      1     0       1       0
69996  99995  22601       1     158   126.0  ...     2      0     0       1       1
69997  99996  19066       2     183   105.0  ...     1      0     1       0       1
69998  99998  22431       1     163    72.0  ...     2      0     0       0       1
69999  99999  20540       1     170    72.0  ...     1      0     0       1 

Cardiovascular Disease Dataset was used for this final project.
This dataset is composed by 13 variables 


#Data description
There are 3 types of input features:

Objective: factual information / this input is easy to measure because either you are male or female or you have age or not.

Examination: results of medical examination / Tells us something that was measured in a process.

Subjective: information given by the patient  /  All the data was collected in reports of patients.



Variables are separated as follows:


VARIABLE | INPUT FEATURES | NAME OF THE VARIABLE IN THE DATASET |  DATA TYPE | (UNIT OF MEASURE ONLY FOR AGE, HEIGHT AND WEIGHT) |  TYPE OF VARIABLE


1. Patient ID | Subjective Feature | ID | int | Quantitative and discrete variable

2. Age | Objective Feature | age | int | (days) | Quantitative and continuous variable

3. Gender | Objective Feature | gender | categorical code | Dummy variable

4. Height | Objective Feature | height | int | (cm) | Quantitative and continuous variable

5. Weight | Objective Feature | weight | float | (kg) | Quantitative and continuous variable

6. Systolic blood pressure | Examination Feature | ap_hi | int | Quantitative and continuous variable

7. Diastolic blood pressure | Examination Feature | ap_lo | int | Quantitative and continuous variable

8. Cholesterol | Examination Feature |  1: normal, 2: above normal, 3: well above normal | Categorical variable

9. Glucose | Examination Feature | gluc | 1: normal, 2: above normal, 3: well above normal | Categorical variable

10. Smoking | Subjective Feature | smoke | binary | Dummy variable

11. Alcohol intake | Subjective Feature | alco | binary | Dummy variable

12. Physical activity | Subjective Feature | active | binary | Dummy variable

13. Presence or absence of cardiovascular disease | Target Variable | cardio | binary | Dummy variable

All of the dataset values were collected at the moment of medical examination.

It is important to highlight that for the dummy variables (smoke, alco, active and cardio -> 0 stands for 'NO' and 1 stands for 'YES'. Talking about the categorical variables ( cholesterol and gluc we only have 3 levels defined as follows: 1 -> normal, 2 -> above normal, 3 -> well above normal.

In [ ]:
Cardio_disease2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           70000 non-null  int64  
 1   age          70000 non-null  int64  
 2   gender       70000 non-null  int64  
 3   height       70000 non-null  int64  
 4   weight       70000 non-null  float64
 5   ap_hi        70000 non-null  int64  
 6   ap_lo        70000 non-null  int64  
 7   cholesterol  70000 non-null  int64  
 8   gluc         70000 non-null  int64  
 9   smoke        70000 non-null  int64  
 10  alco         70000 non-null  int64  
 11  active       70000 non-null  int64  
 12  cardio       70000 non-null  int64  
dtypes: float64(1), int64(12)
memory usage: 6.9 MB


In [ ]:
Cardio_disease2.columns

Index(['id', 'age', 'gender', 'height', 'weight', 'ap_hi', 'ap_lo',
       'cholesterol', 'gluc', 'smoke', 'alco', 'active', 'cardio'],
      dtype='object')

In [ ]:
Cardio_disease2.describe()

id           age  ...        active        cardio
count  70000.000000  70000.000000  ...  70000.000000  70000.000000
mean   49972.419900  19468.865814  ...      0.803729      0.499700
std    28851.302323   2467.251667  ...      0.397179      0.500003
min        0.000000  10798.000000  ...      0.000000      0.000000
25%    25006.750000  17664.000000  ...      1.000000      0.000000
50%    50001.500000  19703.000000  ...      1.000000      0.000000
75%    74889.250000  21327.000000  ...      1.000000      1.000000
max    99999.000000  23713.000000  ...      1.000000      1.000000

[8 rows x 13 columns]

In [ ]:
Cardio_disease2.describe(include='all').T

count          mean           std  ...      50%       75%      max
id           70000.0  49972.419900  28851.302323  ...  50001.5  74889.25  99999.0
age          70000.0  19468.865814   2467.251667  ...  19703.0  21327.00  23713.0
gender       70000.0      1.349571      0.476838  ...      1.0      2.00      2.0
height       70000.0    164.359229      8.210126  ...    165.0    170.00    250.0
weight       70000.0     74.205690     14.395757  ...     72.0     82.00    200.0
ap_hi        70000.0    128.817286    154.011419  ...    120.0    140.00  16020.0
ap_lo        70000.0     96.630414    188.472530  ...     80.0     90.00  11000.0
cholesterol  70000.0      1.366871      0.680250  ...      1.0      2.00      3.0
gluc         70000.0      1.226457      0.572270  ...      1.0      1.00      3.0
smoke        70000.0      0.088129      0.283484  ...      0.0      0.00      1.0
alco         70000.0      0.053771      0.225568  ...      0.0      0.00      1.0
active       70000.0      0.803729      0.397179  ...      1.0      1.00      1.0
cardio       70000.0      0.499700      0.500003  ...      0.0      1.00      1.0

[13 rows x 8 columns]

Looking at the data we can observe some descriptive statistics mean, max, min and dispertion of every single variable. But it is noticeable that some data must be transformed, such as age variable. Because it is expressed in days and  we need to standardize the data therefore it needs to be converted into years. Variables such as ID won't be taken into account. Considering the min and max values we can notice that outliers need to be removed for example: in the case of height there cannot be a person that is only 55 cm tall or the other way around 250 cm tall not even the tallest man in the world is that tall. For Weight some outliers as well do not have to be considered there cant be a person whose weight is 10 kg or even 200 kg. In the following picture we can see the ranges of the Systolic and Diastolic blood pressure. A normal value for Systolic is less than 120 and for Diastolic is 80 therefore negative values cannot be possible besides max values are even 16020 and 11000 accordingly therefore it is evident that some observations known as outliers need to be removed.

> Indented block



# Cholesterol ranges can be found in the image below


In [ ]:
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():

  print('User uploaded file "{name}" with length {length} bytes'.format(

name=fn, length=len(uploaded[fn])))

TypeError: ignored

In [ ]:
import matplotlib.image as mpimg
import matplotlib.pyplot as plt
img = mpimg.imread('image1.png')
plt.imshow(img)

In [ ]:
# Letw take a look about our data and how it looks like the first 4 rows
Cardio_disease2.head()

In [ ]:
Cardio_disease2.dtypes.value_counts()

In [ ]:
#Part about checking the quality fo the data is to know if there are some NOT APPLICABLE values but this dataset has no such observations. Lets create a table where with missing data
miss_values=Cardio_disease2.isnull().sum() # missing values

Perc_values_missing = Cardio_disease2.isnull().sum()/Cardio_disease2.shape[0]*100 # missing value %

value = {
    'miss_values':miss_values,
    'Perc_values_missing':Perc_values_missing 
}
frame=pd.DataFrame(value)
frame

In [ ]:
#Frequency of Dummy and categorical variables
Cardio_disease2[["gender","cholesterol","gluc","smoke","alco","active","cardio" ]].apply(lambda x: x.value_counts()).T.stack()

****************************DATA VISUALIZATION OF THE DATA WITHOUT BEING TRANSFORMED*******************************************************************

In [ ]:
#Let's take a look on the multiple distribution by column
Cardio_disease2.hist(figsize=(20,15));

In [ ]:
# Histograms for continious, discrete, dummy and categorical variables
Cardio_disease2.hist(column=['age', 'gender','height','weight','ap_hi','ap_lo','cholesterol','gluc','smoke','alco','active','cardio'], bins=25, figsize=(20, 15), color='#b4d997', edgecolor='black')

In [ ]:
#CORRELATION MATRIX OF ALL VARIABLES WITHOUT BEING TRANSFORMED
corr=Cardio_disease2.corr()
corr.style.background_gradient(cmap='Oranges_r',axis=None)

After analyzing the dataset we can infere that the best possible option for being our dependent variable is cardio that is a dummy variable. Cardio tell us if there is a presence or absence of cardiovascular disease and 0 stands for absence of cardiovascular disease and 1 stands for presence of cardiovascular disease.

In [ ]:
#Visualizing Dummy and categorical variables and its effects on cardio
Colunas = ('gender', 'cholesterol',	'gluc',	'smoke',	'alco',	'active',	'bmi_high')
fig=plt.figure(figsize=(20,20))
j = 1
for i in Colunas :
  ax=fig.add_subplot(8,8,j)
  j = j + 1
  plt.figure(figsize=(2, 2))
  sns.barplot(x=i,y='cardio',data=Cardio_disease2,ci=None,ax=ax)
  fig.tight_layout()  
plt.show()

********************DATA TRANSFORMATION************************

In [ ]:
#For better understanding about our dummy variable gender that takes values 1 being as female and 2 being as male its better to factorized these values into 0 and 1 where 0 stands for female and 1 as male.
Cardio_disease2['gender_transformed'] = Cardio_disease2['gender'] - 1
Cardio_disease2.head(15)

In [ ]:
#No need to keep original variables of "gender"
del Cardio_disease2['gender']

In [ ]:
# Creation of new Variables
#NEW VARIABLE = age_y
Cardio_disease2['age_years'] = (Cardio_disease2['age']/365).round(decimals=2)    # Age in years -> AGE divided by 365


In [ ]:
sns.histplot(data=Cardio_disease2,x='age_y',bins=30,palette="Paired", alpha=0.9)   

In [ ]:
#2ND NEW VARIABLE
Cardio_disease2['bmi'] = (Cardio_disease2['weight'] / (((Cardio_disease2['height']/100)**2))).round(decimals=2)   # Body mass index


# Creating Flags for categories as 0-1. 0 stands for not having high BMI 1 stands for having high BM1

Cardio_disease2['bmi_high'] = (Cardio_disease2['bmi'] >= 30).astype(int)

In [ ]:
Cardio_disease2.hist(column=['bmi'], bins=50, figsize=(15, 10), color='#b4d997', edgecolor='black')

In [ ]:
# We can observe the total number of patients with high body mass index and people who don't have high BMI
display(pd.DataFrame(Cardio_disease2['bmi_high'].value_counts()))

In [ ]:
print(Cardio_disease2)

In [ ]:
#No need to keep age column since it was transformed already
del Cardio_disease2['age']

In [ ]:
Cardio_disease2.head()

In [ ]:
#As it was mentioned we have outliers that need to be removed from both blood pressures levels Systolic and Diastolic

# First criteria is that we are going to discard the values in which DIastolic is > Systolic because this can not be possible
ap_logreaterthanap_hi = Cardio_disease2[Cardio_disease2['ap_lo'] > Cardio_disease2['ap_hi']].index 

# For the variable ap_hi that is Systolic blood pressure let's drop values higher than 200 or lower than 80
Adjusted_ap_hi = Cardio_disease2[(Cardio_disease2['ap_hi'] > 200) | (Cardio_disease2['ap_hi'] < 80)].index

# For the variable ap_lo that is Diastolic blood pressure let's drop values higher than 160 or lower than 10
Adjusted_ap_lo = Cardio_disease2[(Cardio_disease2['ap_lo'] > 160) | (Cardio_disease2['ap_lo'] < 40)].index

# number of records to be removed
Adjusted_blood = Adjusted_ap_hi.union(Adjusted_ap_lo)
Adjusted_blood.union(Adjusted_ap_lo)
print(Adjusted_blood.size)
1276
# remove the records
Cardio_disease2.drop(Adjusted_blood, inplace = True)


#creating the threshold 0 -> 1 for ap_hi

Cardio_disease2['ap_hi_norm'] = (Cardio_disease2['ap_hi'] < 120).astype(int)
Cardio_disease2['ap_hi_elevated'] = ((Cardio_disease2['ap_hi'] >= 120) & (Cardio_disease2['ap_hi'] <=139)).astype(int)
Cardio_disease2['ap_hi_hypertension']= ((Cardio_disease2['ap_hi'] >=140) & (Cardio_disease2['ap_hi'] <=179)).astype(int)
Cardio_disease2['ap_hi_hyper_crisis']= (Cardio_disease2['ap_hi'] >=180).astype(int)
#creating the threshold 0 -> 1 for ap_lo 

Cardio_disease2['ap_lo_norm'] = (Cardio_disease2['ap_lo'] < 80).astype(int)
Cardio_disease2['ap_lo_elevated'] = ((Cardio_disease2['ap_lo'] >= 80) & (Cardio_disease2['ap_lo'] <=89)).astype(int)
Cardio_disease2['ap_lo_hypertension']= (Cardio_disease2['ap_lo'] >=90).astype(int)
Cardio_disease2['ap_lo_hyper_crisis']= (Cardio_disease2['ap_lo'] >=180).astype(int)
Cardio_disease2

8 new dummy variables were created band the criteria was based on the thresholds that are shown in the image above.

1. ap_hi_norm
2. ap_hi_elevated
3. ap_hi_hypertension
4. ap_hi_hyper_crisis
5. ap_lo_norm
6. ap_lo_elevated
7. ap_lo_hypertension
8. ap_lo_hyper_crisis

In [ ]:
#Lets see some descriptive statistics about our new dataset
display(Cardio_disease2.describe().T)

In [ ]:
# Histograms for dummy variables ao_hi & ap_lo
Cardio_disease2.hist(column=['ap_hi','ap_lo'], bins=25, figsize=(20, 15), color='#b4d997', edgecolor='black')

After the standardization of the data for ap_hi and ap_lo and creations of new dummy variables.
Both variables follows a symmetrical distribution, the mean and the median are close together  if we ordered from smallest to largest the middle value in the data set, resembles the balancing point in the data close to the median and mean. Having a close median to the mean merely indicates the distribution is not skewed too badly. 
Showing us that the data near the mean are more frequent in occurrence than data far from the mean.


****DATA VISUALIZATION OF THE DATA AFTER BEING TRANSFORMED***

In [ ]:
def pie_chart(Cardio_disease2, col, labels):
    data = Cardio_disease2[col].value_counts().to_numpy()
    def absolute_value(val):
        a  = np.round(val / 100 * data.sum(), 0)
        return str('%0.0f' % a) + '\n(' + ('%0.2f' % val) + '%)'
    plt.pie(data, labels = labels, autopct=absolute_value, colors=colors)
    plt.legend(title=col)
    plt.show() 

pie_chart(Cardio_disease2, 'gender_dummy', ['Female','Male'])
pie_chart(Cardio_disease2, 'cardio', ['No','Yes'])
pie_chart(Cardio_disease2, 'cholesterol', ['Normal', 'Above normal', 'Well above normal'])
pie_chart(Cardio_disease2, 'gluc', ['Normal', 'Above normal', 'Well above normal'])
pie_chart(Cardio_disease2, 'smoke', ['No', 'Yes'])
pie_chart(Cardio_disease2, 'alco', ['No','Yes'])
pie_chart(Cardio_disease2, 'active', ['No','Yes'])

#ap_hi 

pie_chart(Cardio_disease2, 'ap_hi_norm', ['No','Yes'])
pie_chart(Cardio_disease2, 'ap_hi_elevated', ['No','Yes'])
pie_chart(Cardio_disease2, 'ap_hi_hypertension', ['No','Yes'])
pie_chart(Cardio_disease2, 'ap_hi_hyper_crisis', ['No','Yes'])

#ap_lo 
pie_chart(Cardio_disease2, 'ap_lo_norm', ['No','Yes'])
pie_chart(Cardio_disease2, 'ap_lo_elevated', ['No','Yes'])
pie_chart(Cardio_disease2, 'ap_lo_hypertension', ['No','Yes'])
pie_chart(Cardio_disease2, 'ap_lo_hyper_crisis', ['No','Yes'])

From the pie charts we can observe that majority of people from the population are female, presence or absence for cardiovascular disease is almost the same. Most of the people with 75% out of the total population have normal cholesterol same as controlled glucose 85% of people are on the normal level of glucose. Very high percentage 91.20% of people don't smoke. Majority of people 94.65% don't consume alcohol but no active 80.34%.
We can also notice there was not level of hyper_crisis just small number in ap_hi and 28% and 25% of people have hypertension and 50% have elevated blood pressure.

In [ ]:
#Scatter plot between weight and age_years based on cardio
plt.figure(figsize=(10, 6))
sns.scatterplot(x='age_years', y='weight', data= Cardio_disease2, hue='cardio')

In [ ]:
#Scatter plot between years and bmi based on cardio
plt.figure(figsize=(15, 7))
sns.scatterplot(x='age_years', y='bmi', data= Cardio_disease2, hue='cardio', legend = "full")

In [ ]:
## Visualizing and understanding continuous variables relation with cardio variable
sns.histplot(Cardio_disease2, x='weight', bins = range(30,150,5), hue='cardio')


In [ ]:
sns.histplot(Cardio_disease2, x='age_years', bins = range(30,70,2), hue='cardio')

In [ ]:
#CORRELATION MAP WITH TRASNFORMED VARIABLES AND NEW CREATED VARIABLES
plt.figure(figsize=(30, 8))
sns.heatmap(Cardio_disease2.corr(), vmin=-1, vmax=1, annot=True, cmap='vlag') # cmap='BrBG'
plt.title('Correlation Map', fontdict={'fontsize':20}, pad=20);

Based on the correlation table we can notice the most significant values, this will help us to define which independents variables we are going to consider for our variable, these are: Cholesterol (0.22), ap_hi_hypertension(0.4), ap_lo_hypertension (0.35), BMI(0.16), age_years(0.24).

In [ ]:
#Creation of objects to define explanatory variables  and dependent variable

x = Cardio_disease2.drop(columns = ['cardio', 'id','age','gender', 'height','weight','ap_hi','ap_lo','gluc','smoke', 'alco','active', 'gender_dummy', 'gender_transformed','ap_hi_norm','ap_hi_at_risk','ap_lo_norm','ap_lo_at_risk', 'ap_hi_elevated', 'ap_hi_hyper_crisis', 'ap_lo_elevated', 'ap_lo_hyper_crisis', 'bmi_high'])

y = Cardio_disease2['cardio']

In [ ]:

# split data
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 8017)



# RobustScaler ALL variables
scaler = RobustScaler()
scaler.fit(x_train)
x_train = pd.DataFrame(scaler.transform(x_train), index=x_train.index, columns=x_train.columns)
x_test = pd.DataFrame(scaler.transform(x_test), index=x_test.index, columns=x_test.columns)

pd.options.display.float_format = '{:,.4f}'.format
display(x_train.head(5))
display(x_train.describe())

GRADIENT BOOSTING

In [ ]:
parameters = {'max_depth': [1,2,3,4,5,6],
              'n_estimators': [200],
              'learning_rate': [0.01]}

# fitting
GradientBoost_GSCV = GridSearchCV(GradientBoostingClassifier(random_state=8017), 
                             parameters, n_jobs=-1, verbose=3)
GradientBoost_GSCV.fit(x_train, y_train)
GradientBoost_model = GradientBoost_GSCV.best_estimator_
GradientBoost_model

In [ ]:
print(GradientBoost_model.get_params())
print(f"Training Score: {round(GradientBoost_model.score(x_train, y_train),4)}")
print(f"Testing Score: {round(GradientBoost_model.score(x_test, y_test),4)}")

LOGISTIC REGRESSION


This statistical model uses a logistic function to model a binary dependent variable. As our dependent variable is has only two values acting like binary  variable with only two classes. Therefore this model was the more convenient.

In [ ]:
#Logistic regression i
Logistic_model =  LogisticRegressionCV(Cs = 50, cv = 5, random_state=8017)
Logistic_model.fit(x_train, y_train)
print(Logistic_model.get_params())

In [ ]:

print('Candidates of Regularization Parameter C:')
print(Logistic_model.Cs_, '\n')

print(f"Training Score: {round(Logistic_model.score(x_train, y_train),4)}")
print(f"Testing Score: {round(Logistic_model.score(x_test, y_test),4)}")

In [ ]:
# Fitted parameters
print(f'Best Regularization Parameter C = {round(Logistic_model.C_[0],4)}')
print(f'intercept = {round(Logistic_model.intercept_[0],4)}')
d = {'estimates' : list(Logistic_model.coef_[0]),
     'absolute' : np.abs(list(Logistic_model.coef_[0]))
    }
table = pd.DataFrame(d, index=x_train.columns)

display(  table.sort_values('absolute', ascending=False).drop(columns='absolute')  )
# Odds ratio of the logistic regression
odds = round(np.exp(table), 3).drop(columns='absolute').sort_values('estimates', ascending=False)
print(odds)

In [ ]:
#Comparison of Models
result = pd.DataFrame({'Model':['Logistic',
                              'Gradient Boost'],
                       'Prediction Accuracy': [Logistic_model.score(x_test, y_test),
                                               GradientBoost_model.score(x_test, y_test),
                                               ]}
                     )
result.sort_values('Prediction Accuracy', ascending=False)

From the 2 models we have got the most accurate Gradient Boost with 0.7219 